File: blog-ws-commons/src/main/scala/io/github/tassiLuca/dse/blog/core/PostsModel.scala
package io.github.tassiLuca.dse.blog.core

import java.util.Date
import scala.util.Try

/** The model of a simple blog posts service. */
trait PostsModel:

  /** The post author's identifier. */
  type AuthorId

  /** The posts title. */
  type Title

  /** The posts body. */
  type Body

  /** The content of the post. */
  type PostContent = (Title, Body)

  /** A post author and their info. */
  case class Author(authorId: AuthorId, name: String, surname: String)

  /** A blog post, comprising an author, title, body and the last modification. */
  case class Post(author: Author, title: Title, body: Body, lastModification: Date)

  /** A function that verifies the content of the post, returning a [[scala.util.Success]] with
    * the content of  the post if the verification succeeds or a [[scala.util.Failure]] otherwise.
    */
  type ContentVerifier = (Title, Body) => Try[PostContent]

  /** A function that verifies the author has appropriate permissions, returning a
    * [[scala.util.Success]] with their information or a [[scala.util.Failure]] otherwise.
    */
  type AuthorsVerifier = AuthorId => Try[Author]

File: blog-ws-direct-kt/src/main/kotlin/io/github/tassiLuca/dse/blog/PostsModel.kt
package io.github.tassiLuca.dse.blog

import java.util.Date

/** The post author's identifier. */
typealias AuthorId = String

/** The post title. */
typealias PostTitle = String

/** The post body. */
typealias PostBody = String

/** The post content, comprising a [title] and a [body]. */
data class PostContent(val title: PostTitle, val body: PostBody)

/** A post author and their info: their [id], [name] and [surname]. */
data class Author(val id: AuthorId, val name: String, val surname: String)

/** A blog post, comprising an [author], its [content] and the [lastModification] date. */
data class Post(
    val author: Author,
    val content: PostContent,
    val lastModification: Date,
)

/** The blog post content verification logic. It returns the post content if it is appropriate, null otherwise. */
typealias ContentVerifier = suspend (PostTitle, PostBody) -> Result<PostContent?>

/** The author verification logic. It returns the author info if it is authorized, null otherwise. */
typealias AuthorsVerifier = suspend (AuthorId) -> Result<Author?>
